pkg/storage/mvcc.go

```diff
@@ -2440,6 +2448,176 @@ func MVCCDeleteRange(
 	return keys, res.ResumeSpan, res.NumKeys, nil
 }
 
+// ExperimentalPredicateMVCCDeleteRange deletes all MVCC versions within the
+// span [key, endKey) which have timestamps in the span (startTime, endTime].
+// This can have the apparent effect of "reverting" the range to startTime,
+// though all older revisions of deleted keys will be GC'ed. Long runs of keys
+// will get deleted with a range Tombstone, while smaller runs
+// will get deleted with point tombstones.
+//
+// Limiting the number of keys or ranges of keys processed can still cause a
+// batch that is too large -- in number of bytes -- for raft to replicate if the
+// keys are very large. So if the total length of the keys or key spans cleared
+// exceeds maxBatchByteSize it will also stop and return a resume span.
+//
+// If the underlying iterator encounters an intent with a timestamp in the span
```

Comment:
should we scan for intents/inline meta before we start issuing tombstones, or should we return a resume span once `ExperimentalMVCCDeleteRangeUsingTombstone` or `MVCCDeleteRange` encounters one?

Reply:
Scan for intents first, using e.g. `ScanIntents`. Now that we have separated intents that's a very cheap operation -- essentially free in the typical case of no intents. That'll also allow returning a bunch of intents for resolution at once, which improves performance when we find stale intents.

Reply:
Sounds good.